File: java/SharedQueue.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Thread-safe shared queue implementation using ReentrantLock and Condition
 */
public class SharedQueue {
    private final Queue<Task> queue;
    private final ReentrantLock lock;
    private final Condition notEmpty;
    private final Condition notFull;
    private final int maxSize;
    private volatile boolean shutdown = false;
    
    public SharedQueue(int maxSize) {
        this.maxSize = maxSize;
        this.queue = new LinkedList<>();
        this.lock = new ReentrantLock();
        this.notEmpty = lock.newCondition();
        this.notFull = lock.newCondition();
    }
    
    /**
     * Add a task to the queue
     * @param task The task to add
     * @return true if added successfully, false if queue is full or shutdown
     */
    public boolean addTask(Task task) {
        if (task == null) {
            throw new IllegalArgumentException("Task cannot be null");
        }
        
        lock.lock();
        try {
            while (queue.size() >= maxSize && !shutdown) {
                try {
                    notFull.await();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return false;
                }
            }
            
            if (shutdown) {
                return false;
            }
            
            queue.offer(task);
            notEmpty.signal();
            return true;
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Get a task from the queue
     * @return The task, or null if queue is empty and shutdown
     */
    public Task getTask() {
        lock.lock();
        try {
            while (queue.isEmpty() && !shutdown) {
                try {
                    notEmpty.await();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return null;
                }
            }
            
            if (queue.isEmpty() && shutdown) {
                return null;
            }
            
            Task task = queue.poll();
            notFull.signal();
            return task;
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Check if queue is empty
     * @return true if empty, false otherwise
     */
    public boolean isEmpty() {
        lock.lock();
        try {
            return queue.isEmpty();
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Get current queue size
     * @return current size
     */
    public int size() {
        lock.lock();
        try {
            return queue.size();
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Shutdown the queue
     */
    public void shutdown() {
        lock.lock();
        try {
            shutdown = true;
            notEmpty.signalAll();
            notFull.signalAll();
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Check if queue is shutdown
     * @return true if shutdown, false otherwise
     */
    public boolean isShutdown() {
        return shutdown;
    }
}

File: go/shared_queue.go
package main

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// SharedQueue represents a thread-safe queue for tasks
type SharedQueue struct {
	tasks     chan *Task
	mu        sync.RWMutex
	shutdown  bool
	maxSize   int
	ctx       context.Context
	cancel    context.CancelFunc
}

// NewSharedQueue creates a new shared queue with the specified max size
func NewSharedQueue(maxSize int) *SharedQueue {
	ctx, cancel := context.WithCancel(context.Background())
	return &SharedQueue{
		tasks:    make(chan *Task, maxSize),
		maxSize:  maxSize,
		ctx:      ctx,
		cancel:   cancel,
		shutdown: false,
	}
}

// AddTask adds a task to the queue
func (sq *SharedQueue) AddTask(task *Task) bool {
	if task == nil {
		return false
	}

	sq.mu.RLock()
	if sq.shutdown {
		sq.mu.RUnlock()
		return false
	}
	sq.mu.RUnlock()

	select {
	case sq.tasks <- task:
		return true
	case <-sq.ctx.Done():
		return false
	default:
		// Queue is full, try to add with timeout
		select {
		case sq.tasks <- task:
			return true
		case <-time.After(100 * time.Millisecond):
			return false
		case <-sq.ctx.Done():
			return false
		}
	}
}

// GetTask retrieves a task from the queue
func (sq *SharedQueue) GetTask() *Task {
	sq.mu.RLock()
	if sq.shutdown {
		sq.mu.RUnlock()
		return nil
	}
	sq.mu.RUnlock()

	select {
	case task := <-sq.tasks:
		return task
	case <-sq.ctx.Done():
		return nil
	}
}

// IsEmpty checks if the queue is empty
func (sq *SharedQueue) IsEmpty() bool {
	sq.mu.RLock()
	defer sq.mu.RUnlock()
	return len(sq.tasks) == 0
}

// Size returns the current size of the queue
func (sq *SharedQueue) Size() int {
	sq.mu.RLock()
	defer sq.mu.RUnlock()
	return len(sq.tasks)
}

// Shutdown shuts down the queue
func (sq *SharedQueue) Shutdown() {
	sq.mu.Lock()
	defer sq.mu.Unlock()
	if !sq.shutdown {
		sq.shutdown = true
		sq.cancel()
		close(sq.tasks)
	}
}

// IsShutdown checks if the queue is shutdown
func (sq *SharedQueue) IsShutdown() bool {
	sq.mu.RLock()
	defer sq.mu.RUnlock()
	return sq.shutdown
}

// String returns a string representation of the queue status
func (sq *SharedQueue) String() string {
	sq.mu.RLock()
	defer sq.mu.RUnlock()
	return fmt.Sprintf("SharedQueue{size=%d, maxSize=%d, shutdown=%v}", len(sq.tasks), sq.maxSize, sq.shutdown)
}
